### ISSUE TITLE
MemoryBody blows up with "java.nio.ReadOnlyBufferException" if constructed with read-only ByteBuffer

### ISSUE THREAD

**Issue by xenomachina** [http4k/http4k#233]
This test runs successfully:

```kotlin
    @Test
    fun `MemoryBody works with ByteBuffer`() {
        val bb = ByteBuffer.wrap("Goodbye".toByteArray())
        val body = Body(bb)
        assertThat(body.length, equalTo(7L))
        assertThat(body.toString(), equalTo("Goodbye"))
        assertThat(body.stream.bufferedReader().readText(), equalTo("Goodbye"))
    }
```

however, this variant fails:

```kotlin
    @Test
    fun `MemoryBody works with read-only ByteBuffer`() {
        val bb = ByteBuffer.wrap("Goodbye".toByteArray())
        val body = Body(bb.asReadOnlyBuffer())
        assertThat(body.length, equalTo(7L))
        assertThat(body.toString(), equalTo("Goodbye"))
        assertThat(body.stream.bufferedReader().readText(), equalTo("Goodbye"))
    }
```

Despite the fact that `MemoryBody` only "reads" from the `ByteBuffer`, it does this by getting its array, which is not allowed on read-only `ByteBuffer` instances.

**Comment by xenomachina** [comment]
To clarify, construction of the `Body` (actually `MemoryBody`) succeeds, but calling any of its methods other than `close` will cause `ReadOnlyBufferException` to be thrown.

**Comment by daviddenton** [comment]
Thanks for spotting this @xenomachina . I've got a speculative fix in this branch: https://github.com/http4k/http4k/pull/234/commits which would treat non-array backed ByteBuffers as streams.

**Comment by xenomachina** [comment]
Thanks for looking into this!

One downside to treating them as streams is that they can't be re-read. A read-only `ByteBuffer` can be read multiple times. I assume the reason the array itself is not accessible is because there's no such thing as a read-only array on the JVM.

**Comment by daviddenton** [comment]
@xenomachina So I've updated the PR with the alternative, which is to read the bytes into an array and then wrap that in another ByteBuffer which isn't read only. The new version has the advantage that we avoid implementing the inputstream, which personally I prefer. I'd like to keep the implementation as simple as possible for the moment, so I'm just going to merge the new version in for the time being since it does mitigate the bug - guessing that if people really want to use the stream version then we can probably push them to implement it using the same approach we took with the custom InputStream. But thoughts welcome as always :)

Curious as to the usecase for this - how did you come across it?

**Comment by xenomachina** [comment]
Thanks, @daviddenton!

We've got some code that creates a simulated `Response` object based on values we're getting from AWS. We recently upgraded from [version 1 of the AWS SDK for Java](https://github.com/aws/aws-sdk-java) to [version 2 of the AWS SDK for Java](https://github.com/aws/aws-sdk-java-v2). The old version of the API gave us back a `ByteBuffer` containing the bytes from our `Body`. Apparently it was a writable `ByteBuffer`, and so it behaved as expected. 

The new version of this API gives us an [`SdkBytes` object](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/core/SdkBytes.html). That object in turn has methods to convert it into a byte array, `ByteBuffer`, or `InputStream`. In migrating to the new the API we just took the path of least resistance, and used the `asByteBuffer()` method in place of the `ByteBuffer` it used to give us. Our tests then blew up whenever they tried to use the constructed `Body` objects, as the `ByteBuffer` returned by `SdkBytes` is read-only.

I can create a `Body` implementation that uses `SdkBytes`, but I thought you might want to know about this behavior with read-only ByteBuffers, as it seems quite surprising. It's too bad there isn't a standard class in Java or Kotlin for a read-only array of bytes. It's kind of crazy that there isn't even a `ByteBufferInputStream` in the standard library.
